packages/flutter_tools/lib/src/runner/flutter_command_runner.dart

```diff
@@ -242,9 +245,59 @@ class FlutterCommandRunner extends CommandRunner<void> {
       }
     }
 
+    _machineFlagPresentInAnyCliArg = args.contains('--${FlutterGlobalOptions.kMachineFlag}');
     return super.run(args);
   }
 
+  /// Whether to perform a flutter version check, which prints a warning if old.
+  Future<bool> _shouldCheckForUpdates(
+    ArgResults topLevelResults, {
+    required bool topLevelMachineFlag,
+  }) async {
+    // If the top level --machine flag is set, we don't want to check for updates.
+    if (topLevelMachineFlag) {
+      return false;
+    }
+
+    // If the same flag appears in any subcommand, we don't want to check for updates.
+    //
+    // A better solution would be the flag not being in any specific subcommand, just
+    // in the top level command, but that would require a more significant refactor
+    // and deprecation of the current behaviors.
+    //
+    // See https://github.com/flutter/flutter/issues/145158.
+    if (_machineFlagPresentInAnyCliArg) {
+      return false;
+    }
+
+    // Running the "upgrade" command is already checking, don't check twice.
+    if (topLevelResults.command?.name == 'upgrade') {
+      return false;
+    }
+
+    // Check if the user has explicitly requested a version check.
+    final bool versionCheckFlag = topLevelResults[FlutterGlobalOptions.kVersionCheckFlag] as bool? ?? false;
+    final bool explicitVersionCheckPassed = topLevelResults.wasParsed(FlutterGlobalOptions.kVersionCheckFlag) && versionCheckFlag;
+    if (explicitVersionCheckPassed) {
+      return true;
+    }
+
+    // Implicitly suppress if any of the following are true:
+    // 1. We are not attached to a terminal.
+    // 2. The top level command appears to be related to bash/zsh completion.
+    // 3. We are running in a known CI environment.
+    final bool notAttachedToTerminal = !globals.stdio.hasTerminal;
+    final bool redirectedCompletion = (topLevelResults.command?.name ?? '').endsWith('-completion');
+    if (notAttachedToTerminal ||
+        redirectedCompletion ||
+        await globals.botDetector.isRunningOnBot) {
+      return false;
+    }
```

Comment:
Nit: does this comment provide value? To me, it seems redundant with the code (though `redirectedCompletion` seems like an odd name to me.
```suggestion
    final bool notAttachedToTerminal = !globals.stdio.hasTerminal;
    // e.g. `flutter bash-completion` or `flutter zsh-completion`
    final bool isShellCompletionCommand = (topLevelResults.command?.name ?? '').endsWith('-completion');
    if (notAttachedToTerminal ||
        isShellCompletionCommand ||
        await globals.botDetector.isRunningOnBot) {
      return false;
    }
```